pytorch_lightning/callbacks/finetuning.py

```diff
@@ -17,166 +17,262 @@
+    This class implements the base logic for writing your own Finetuning Callback.
 
-    @staticmethod
-    def _make_trainable(module: Module) -> None:
-        """Unfreezes a given module.
-        Args:
-            module: The module to unfreeze
-        """
-        for param in module.parameters():
-            param.requires_grad = True
-        module.train()
+    Override ``freeze_before_training`` and ``finetune_function`` methods with your own logic.
+
+    ``freeze_before_training``: This method is called before ``configure_optimizers``
+        and should be used to freeze any modules parameters.
+
+    ``finetune_function``: This method is called on every train epoch start and should be used to
+        ``unfreeze`` any parameters. Those parameters needs to be added in a new ``param_group``
+        within the optimizer.
+
+    .. note:: Make sure to filter the parameters based on `requires_grad`.
+
+    Example::
+
+        class MyModel(LightningModule)
+
+            ...
+
+            def configure_optimizer(self):
+                # Make sure to filter the parameters based on `requires_grad`
+                return Adam(filter(lambda p: p.requires_grad, self.parameters))
+
+        class FeatureExtractorFreezeUnfreeze(BaseFinetuning):
+
+            def __init__(self, unfreeze_at_epoch=10)
+                self._unfreeze_at_epoch = unfreeze_at_epoch
+
+            def freeze_before_training(self, pl_module):
+                # freeze any module you want
+                # Here, we are freezing ``feature_extractor``
+                self.freeze(pl_module.feature_extractor)
+
+            def finetune_function(self, pl_module, current_epoch, optimizer, optimizer_idx):
+                # When ``current_epoch`` is 10, feature_extractor will start training.
+                if current_epoch == self._unfreeze_at_epoch:
+                    self.unfreeze_and_add_param_group(
+                        module=pl_module.feature_extractor,
+                        optimizer=optimizer,
+                        train_bn=True,
+                    )
+    """
 
     @staticmethod
-    def _recursive_freeze(module: Module,
-                          train_bn: bool = True) -> None:
-        """Freezes the layers of a given module.
-        Args:
-            module: The module to freeze
-            train_bn: If True, leave the BatchNorm layers in training mode
-        """
-        children = list(module.children())
-        if not children:
-            if not (isinstance(module, BaseFinetuningCallback.BN_TYPES) and train_bn):
-                for param in module.parameters():
-                    param.requires_grad = False
-                module.eval()
-            else:
-                # Make the BN layers trainable
-                BaseFinetuningCallback._make_trainable(module)
+    def flatten_modules(modules: Union[Module, List[Module]]) -> List[Module]:
+        if isinstance(modules, list):
```

Comment:
```suggestion
        if isinstance(modules, (tuple, list)):
```
or rather any iterable?

Reply:
Good idea !